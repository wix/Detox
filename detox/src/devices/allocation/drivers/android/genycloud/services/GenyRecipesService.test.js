describe('Genymotion-Cloud recipes service', () => {
  const givenNoRecipes = () => {
    exec.getRecipe.mockResolvedValue({
      recipes: [],
    });
  };

  const givenRecipes = (...recipes) => {
    exec.getRecipe.mockResolvedValue({
      recipes: [...recipes],
    });
  };

  const aRecipe = () => ({
    uuid: 'mock-recipe-uuid',
    name: 'mock-recipe-name',
  });

  const anotherRecipe = () => ({
    uuid: 'another-mock-recipe-uuid',
    name: 'another-mock-recipe-name',
  });

  let exec;
  let uut;
  beforeEach(() => {
    const GenyCloudExec = jest.createMockFromModule('../exec/GenyCloudExec');
    exec = new GenyCloudExec();

    const GenyRecipesService = require('./GenyRecipesService');
    uut = new GenyRecipesService(exec);
  });

  describe('getting a recipe by name', () => {
    it('should throw an error if no recipes found', async () => {
      givenNoRecipes();
      await expect(uut.getRecipeByName('mock-name')).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should return the recipe if exactly one match is found', async () => {
      const recipe = aRecipe();
      givenRecipes(recipe);

      const result = await uut.getRecipeByName(recipe.name);

      expect(result.uuid).toEqual(recipe.uuid);
    });

    it('should return a recipe DTO', async () => {
      const recipe = aRecipe();
      givenRecipes(recipe);

      const result = await uut.getRecipeByName(recipe.name);
      expect(result.constructor.name).toContain('Recipe');
    });

    it('should throw an error if there are multiple matches', async () => {
      const recipe = aRecipe();
      const recipe2 = anotherRecipe();
      givenRecipes(recipe, recipe2);

      await expect(uut.getRecipeByName(recipe.name)).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('Getting a recipe by UUID', () => {
    it('should immediately return a recipe', async () => {
      const recipe = aRecipe();

      const result = await uut.getRecipeByUUID(recipe.uuid);

      expect(result.uuid).toEqual(recipe.uuid);
      expect(result.constructor.name).toContain('Recipe');
    });
  });
});
