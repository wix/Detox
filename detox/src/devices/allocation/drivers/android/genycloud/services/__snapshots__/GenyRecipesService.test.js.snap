// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Genymotion-Cloud recipes service getting a recipe by name should throw an error if no recipes found 1`] = `
"No Genymotion-Cloud recipe found for recipe name "mock-name"

HINT: Please check your recipe name or use recipe UUID instead."
`;

exports[`Genymotion-Cloud recipes service getting a recipe by name should throw an error if there are multiple matches 1`] = `
"More than one Genymotion-Cloud recipe found for recipe name mock-recipe-name:
  mock-recipe-name (mock-recipe-uuid)
  another-mock-recipe-name (another-mock-recipe-uuid)

HINT: Please specify a unique recipe name or use recipe UUID instead."
`;
